File: DotNetGrc/Grc/Ast/Node/NodeBase.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Grc.Ast.Visitor;

namespace Grc.Ast.Node
{
	public abstract partial class NodeBase
	{
		private string text;

		public string Text { get { return text ?? (text = GetText()); } }

		public abstract int Line { get; }

		public abstract int Pos { get; }

		public string Location
		{
			get { return string.Format("[{0}, {1}]", Line, Pos); }
		}

		public NodeBase Parent { get; set; }

		protected virtual int Indent
		{
			get { return Parent != null ? Parent.Indent + 1 : 0; }
		}

		protected string Tabs
		{
			get
			{
				StringBuilder sb = new StringBuilder();

				for (int i = 0; i < Indent; i++)
					sb.Append("\t");

				return sb.ToString();
			}
		}

		public abstract void Accept(IVisitor v);

		protected abstract string GetText();

		public override string ToString()
		{
			return Text;
		}
	}
}

File: Parser/src/main/java/k31/grc/ast/node/NodeBase.java
package k31.grc.ast.node;

import java.util.LinkedList;
import java.util.List;

import k31.grc.ast.visitor.Visitor;
import k31.grc.cst.visitor.GVNode;

public abstract class NodeBase {

	private NodeBase parent;

	private List<NodeBase> children;

	private String text;

	public NodeBase getParent() {

		return parent;
	}

	public List<NodeBase> getChildren() {

		return children;
	}

	public String getText() {

		return text;
	}

	public NodeBase(String text) {

		this.text = text;

		this.children = new LinkedList<NodeBase>();
	}

	public void addChild(NodeBase c) {

		c.parent = this;

		children.add(c);
	}

	public abstract void accept(Visitor v);

	@Override
	public String toString() {

		return text;
	}
}
